packages/table/src/headers/header.tsx

```diff
@@ -224,35 +225,20 @@ const SHALLOW_COMPARE_PROP_KEYS_BLACKLIST: Array<keyof IInternalHeaderProps> = [
 const RESET_CELL_KEYS_BLACKLIST: Array<keyof IInternalHeaderProps> = ["indexEnd", "indexStart"];
 
 export class Header extends React.Component<IInternalHeaderProps, IHeaderState> {
-    public state: IHeaderState = {
-        hasSelectionEnded: false,
-    };
-
     protected activationIndex: number;
     private batcher = new Batcher<JSX.Element>();
 
-    public constructor(props?: IHeaderProps, context?: any) {
+    public constructor(props?: IInternalHeaderProps, context?: any) {
         super(props, context);
-    }
-
-    public componentDidMount() {
-        if (this.props.selectedRegions != null && this.props.selectedRegions.length > 0) {
-            // we already have a selection defined, so we'll want to enable reordering interactions
-            // right away if other criteria are satisfied too.
-            this.setState({ hasSelectionEnded: true });
-        }
+        this.state = { hasSelectionEnded: this.isSelectedRegionsControlledAndNonEmpty(props) };
     }
 
     public componentWillUnmount() {
         this.batcher.cancelOutstandingCallback();
     }
 
     public componentWillReceiveProps(nextProps?: IInternalHeaderProps) {
-        if (nextProps.selectedRegions != null && nextProps.selectedRegions.length > 0) {
-            this.setState({ hasSelectionEnded: true });
-        } else {
-            this.setState({ hasSelectionEnded: false });
-        }
+        this.setState({ hasSelectionEnded: this.isSelectedRegionsControlledAndNonEmpty(nextProps) });
```

Comment:
⚠️ this will always update the state when props change; was that your intention behind this change? or does it need to be gated with an `if` clause?

Reply:
Yep, that's intended. For row headers, `hasSelectionEnded` needs to be `true` before reordering is allowed. When you pass in `selectedRegions` in a controlled fashion, we want the flag to be true from the get-go so that reordering is immediately enabled.

Reply:
Changed to `hasValidSelection`.